istio-interop-testing/src/test/java/io/grpc/testing/istio/EchoTestServerTest.java

```diff
@@ -50,6 +60,30 @@
 @RunWith(JUnit4.class)
 public class EchoTestServerTest {
 
+  private static final String[] EXPECTED_KEY_SET = {
+      "--server_first", "--forwarding-address",
+      "--bind_ip", "--istio-version", "--bind_localhost", "--grpc", "--tls",
+      "--cluster", "--key", "--tcp", "--crt", "--metrics", "--port", "--version"
+  };
+
+  private static final String TEST_ARGS =
+      "--metrics=15014 --cluster=\"cluster-0\" --port=\"18080\" --grpc=\"17070\" --port=\"18085\""
+          + " --tcp=\"19090\" --port=\"18443\" --tls=18443 --tcp=\"16060\" --server_first=16060"
+          + " --tcp=\"19091\" --tcp=\"16061\" --server_first=16061 --port=\"18081\""
+          + " --grpc=\"17071\" --port=\"19443\" --tls=19443 --port=\"18082\" --bind_ip=18082"
+          + " --port=\"18084\" --bind_localhost=18084 --tcp=\"19092\" --port=\"18083\""
+          + " --port=\"8080\" --port=\"3333\" --version=\"v1\" --istio-version=3 --crt=/cert.crt"
+          + " --key=/cert.key --forwarding-address=192.168.1.10:7072";
+
+  private static final String TEST_ARGS_PORTS =
+      "--metrics=15014 --cluster=\"cluster-0\" --port=\"18080\" --grpc=17070 --port=18085"
+          + " --tcp=\"19090\" --port=\"18443\" --tls=18443 --tcp=16060 --server_first=16060"
+          + " --tcp=\"19091\" --tcp=\"16061\" --server_first=16061 --port=\"18081\""
+          + " --grpc=\"17071\" --port=\"19443\" --tls=\"19443\" --port=\"18082\" --bind_ip=18082"
+          + " --port=\"18084\" --bind_localhost=18084 --tcp=\"19092\" --port=\"18083\""
+          + " --port=\"8080\" --port=3333 --version=\"v1\" --istio-version=3 --crt=/cert.crt"
```

Comment:
Just one note about the ports and quotes. I was thinking about why we even need to handle them separately.
I don't think having quotes in these test examples makes sense. IRRC shells won't pass them, unless the user _specifically_ escapes the quote signs.

```
❯ ./args.sh arg1 --port=1 --port="123" --port="\"42\""
Script name is: ./args.sh
Arg 1 is arg1
Arg 2 is --port=1
Arg 3 is --port=123
Arg 4 is --port="42"
-----------
All args: arg1 --port=1 --port=123 --port="42"
All args count: 4
```

The file I used was
```sh
#!/bin/bash

echo "Script name is: $0"
echo "Arg 1 is $1"
echo "Arg 2 is $2"
echo "Arg 3 is $3"
echo "Arg 4 is $4"
echo "-----------"
echo "All args: $*"
echo "All args count: $#"
```

I recommend removing any quotes from these test examples, from the expected results of assertions, and to remove the extra logic in the port parsing logic.

Reply:
True when the shell is processing the command line. But if a program is directly invoking a child program the quotes are not stripped.

I did a test:

Child process:
```
package io.grpc.gcp.observability;

public class ArgsTrial {
    public static void main(String[] args) {
        for (String arg: args) {
            System.out.println("next arg:" + arg);
        }
    }
}
```
 
Parent process:
```
package io.grpc.gcp.observability;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ParentTrial {
    public static void main(String[] args) throws IOException, InterruptedException {
        List<String> myArgs = Arrays.asList("arg1", "arg2=\"sub\"");
        System.out.println("My args are:" + myArgs);
        System.out.println("retcode=" + exec("io.grpc.gcp.observability.ArgsTrial",
                Collections.emptyList(), myArgs));
    }

    public static int exec(String className, List<String> jvmArgs, List<String> args) throws IOException,
            InterruptedException {
        String javaHome = System.getProperty("java.home");
        String javaBin = javaHome + File.separator + "bin" + File.separator + "java";
        String classpath = System.getProperty("java.class.path");

        List<String> command = new ArrayList<>();
        command.add(javaBin);
        command.addAll(jvmArgs);
        command.add("-cp");
        command.add(classpath);
        command.add(className);
        command.addAll(args);

        ProcessBuilder builder = new ProcessBuilder(command);
        Process process = builder.inheritIO().start();
        process.waitFor();
        return process.exitValue();
    }
}
```

Generate the child class:

```
javac io/grpc/gcp/observability/ArgsTrial.java
```

Execute the test:
```
java io/grpc/gcp/observability/ParentTrial.java 
My args are:[arg1, arg2="sub"]
next arg:arg1
next arg:arg2="sub"
retcode=0
```

As you can see the quotes are *not* stripped in the arg passed to the child process.

Not sure why I added this though - I thought the original one (C++ echo server) had it but I don't see it now. In any case if the program is invoked not via a shell command then this logic will be useful.